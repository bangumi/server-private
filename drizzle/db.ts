import type { ExtractTablesWithRelations } from 'drizzle-orm';
import type { MySqlTransaction } from 'drizzle-orm/mysql-core';
import {
  drizzle,
  type MySql2PreparedQueryHKT,
  type MySql2QueryResultHKT,
} from 'drizzle-orm/mysql2';
import * as mysql from 'mysql2/promise';

import config, { developing, testing } from '@app/lib/config.ts';
import { logger } from '@app/lib/logger.ts';

import * as schema from './schema.ts';

const poolConnection = mysql.createPool({
  host: config.mysql.host,
  user: config.mysql.user,
  database: config.mysql.db,
  password: config.mysql.password,
  port: config.mysql.port,
});

export const db = drizzle(poolConnection, {
  schema,
  mode: 'default',
  logger:
    testing || developing
      ? {
          logQuery(query: string, params: unknown[]) {
            logger.trace('query', { query, params });
          },
        }
      : undefined,
});

export * as op from 'drizzle-orm';

export type Txn = MySqlTransaction<
  MySql2QueryResultHKT,
  MySql2PreparedQueryHKT,
  typeof schema,
  ExtractTablesWithRelations<Record<string, unknown>>
>;
