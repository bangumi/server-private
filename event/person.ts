import { getSlimCacheKey } from '@app/lib/person/cache';
import redis from '@app/lib/redis.ts';

import { EventOp } from './type';

interface PersonKey {
  prsn_id: number;
}

interface Payload {
  op: EventOp;
}

export async function handle(topic: string, key: string, value: string) {
  const idx = JSON.parse(key) as PersonKey;
  const payload = JSON.parse(value) as Payload;
  switch (payload.op) {
    case EventOp.Create: {
      break;
    }
    case EventOp.Update:
    case EventOp.Delete: {
      await redis.del(getSlimCacheKey(idx.prsn_id));
      break;
    }
    case EventOp.Snapshot: {
      break;
    }
  }
}
