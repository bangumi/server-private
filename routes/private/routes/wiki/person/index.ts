import type { Static } from '@sinclair/typebox';
import { Type as t } from '@sinclair/typebox';

import { NotAllowedError } from '@app/lib/auth/index.ts';
import { BadRequestError, NotFoundError } from '@app/lib/error.ts';
import { Security, Tag } from '@app/lib/openapi/index.ts';
import type { PersonRev } from '@app/lib/orm/entity/index.ts';
import { createRevision, RevType } from '@app/lib/orm/entity/index.ts';
import { AppDataSource, entity, PersonRepo } from '@app/lib/orm/index.ts';
import { InvalidWikiSyntaxError } from '@app/lib/subject/index.ts';
import * as req from '@app/lib/types/req.ts';
import * as res from '@app/lib/types/res.ts';
import { formatErrors } from '@app/lib/types/res.ts';
import { matchExpected, WikiChangedError } from '@app/lib/wiki.ts';
import { requireLogin } from '@app/routes/hooks/pre-handler.ts';
import type { App } from '@app/routes/type.ts';

export const PersonWikiInfo = t.Object(
  {
    id: t.Integer(),
    name: t.String(),
    typeID: t.Ref(req.SubjectType),
    infobox: t.String(),
    summary: t.String(),
  },
  { $id: 'PersonWikiInfo' },
);

export const PersonEdit = t.Object(
  {
    name: t.String({ minLength: 1 }),
    infobox: t.String({ minLength: 1 }),
    summary: t.String(),
  },
  {
    $id: 'PersonEdit',
    additionalProperties: false,
  },
);

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.addSchema(PersonWikiInfo);

  app.get(
    '/persons/:personID',
    {
      schema: {
        tags: [Tag.Wiki],
        operationId: 'getPersonWikiInfo',
        description: '获取当前的 wiki 信息',
        params: t.Object({
          personID: t.Integer({ examples: [1], minimum: 0 }),
        }),
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        response: {
          200: t.Ref(PersonWikiInfo),
          401: t.Ref(res.Error, {
            'x-examples': formatErrors(new InvalidWikiSyntaxError()),
          }),
          404: t.Ref(res.Error, {
            description: '角色不存在',
          }),
        },
      },
    },
    async ({ params: { personID } }): Promise<Static<typeof PersonWikiInfo>> => {
      const p = await PersonRepo.findOneBy({ id: personID, redirect: 0 });
      if (!p) {
        throw new NotFoundError(`person ${personID}`);
      }

      if (p.lock) {
        throw new NotAllowedError('edit a locked person');
      }

      return {
        id: p.id,
        name: p.name,
        infobox: p.infobox,
        summary: p.summary,
        typeID: p.type,
      };
    },
  );

  app.patch(
    '/persons/:personID',
    {
      schema: {
        tags: [Tag.Wiki],
        operationId: 'patchPersonInfo',
        params: t.Object({
          personID: t.Integer({ examples: [1], minimum: 0 }),
        }),
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        body: t.Object(
          {
            commitMessage: t.String({ minLength: 1 }),
            expectedRevision: t.Partial(PersonEdit, { default: {}, additionalProperties: false }),
            person: t.Partial(PersonEdit, { additionalProperties: false }),
          },
          { additionalProperties: false },
        ),
        response: {
          200: t.Object({}),
          400: t.Ref(res.Error, {
            'x-examples': formatErrors(new WikiChangedError('name', '1', '2')),
          }),
          401: t.Ref(res.Error, {
            'x-examples': formatErrors(new InvalidWikiSyntaxError()),
          }),
        },
      },
      preHandler: [requireLogin('editing a subject info')],
    },
    async ({
      auth,
      body: { commitMessage, person: input, expectedRevision },
      params: { personID },
    }) => {
      if (!auth.permission.mono_edit) {
        throw new NotAllowedError('edit person');
      }

      await AppDataSource.transaction(async (t) => {
        const PersonRepo = t.getRepository(entity.Person);
        const p = await PersonRepo.findOneBy({ id: personID });
        if (!p) {
          throw new NotFoundError(`person ${personID}`);
        }
        if (p.lock || p.redirect) {
          throw new BadRequestError('locked person');
        }

        matchExpected(p, expectedRevision);

        p.infobox = input.infobox ?? p.infobox;
        p.name = input.name ?? p.name;
        p.summary = input.summary ?? p.summary;

        await PersonRepo.save(p);

        await createRevision(t, {
          mid: personID,
          type: RevType.personEdit,
          rev: {
            crt_name: p.name,
            crt_infobox: p.infobox,
            crt_summary: p.summary,
            extra: {
              img: p.img,
            },
          } satisfies PersonRev,
          creator: auth.userID,
          comment: commitMessage,
        });
      });

      return {};
    },
  );
}
