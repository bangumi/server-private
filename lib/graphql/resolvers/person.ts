import type { Wiki } from '@bgm38/wiki';
import { parse as parseWiki, WikiSyntaxError } from '@bgm38/wiki';

import type * as entity from '@app/lib/orm/entity/index.ts';
import { personImages } from '@app/lib/response.ts';

function convertCareer(person: entity.Person) {
  const result: string[] = [];
  if (person.producer) {
    result.push('producer');
  }
  if (person.mangaka) {
    result.push('mangaka');
  }
  if (person.artist) {
    result.push('artist');
  }
  if (person.seiyu) {
    result.push('seiyu');
  }
  if (person.writer) {
    result.push('writer');
  }
  if (person.illustrator) {
    result.push('illustrator');
  }
  if (person.actor) {
    result.push('actor');
  }
  return result;
}

export function convertPerson(person: entity.Person) {
  let wiki: Wiki = {
    type: '',
    data: [],
  };
  try {
    wiki = parseWiki(person.infobox);
  } catch (error) {
    if (!(error instanceof WikiSyntaxError)) {
      throw error;
    }
  }
  const infobox = wiki.data.map((item) => {
    if (item.array) {
      return item;
    }
    return {
      key: item.key,
      values: [
        {
          v: item.value,
        },
      ],
    };
  });
  return {
    id: person.id,
    name: person.name,
    type: person.type,
    infobox,
    career: convertCareer(person),
    summary: person.summary,
    images: personImages(person.img),
    comment: person.comment,
    collects: person.collects,
    last_post: person.lastPost,
    lock: person.lock,
    redirect: person.redirect,
    nsfw: person.nsfw,
  };
}
