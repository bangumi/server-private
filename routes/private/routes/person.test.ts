import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import { emptyAuth } from '@app/lib/auth/index.ts';
import { db, op } from '@app/drizzle/db.ts';
import * as schema from '@app/drizzle/schema.ts';
import redis from '@app/lib/redis.ts';
import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './person.ts';

describe('person', () => {
  test('should get person', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/persons/1',
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get person works', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/persons/1/works',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get person casts', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/persons/1/casts',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get person collects', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/persons/1/collects',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });
});

describe('person comments', () => {
  beforeEach(async () => {
    await redis.flushdb();
    await db.delete(schema.chiiPrsnComments).where(op.eq(schema.chiiPrsnComments.mid, 1));
    await db.insert(schema.chiiPrsnComments).values({
      id: 12345670,
      mid: 1,
      content: '7月7日 77结婚',
      state: 0,
      createdAt: 1400517144,
      uid: 287622,
      related: 0,
    });
    await db.insert(schema.chiiPrsnComments).values({
      id: 12345671,
      mid: 1,
      content: '是6号结的',
      state: 0,
      createdAt: 1400517144,
      uid: 287622,
      related: 12345670,
    });
    await db.insert(schema.chiiPrsnComments).values({
      id: 12345672,
      mid: 1,
      content: '生日快乐！',
      state: 0,
      createdAt: 1400517144,
      uid: 287622,
      related: 0,
    });
  });

  afterEach(async () => {
    await redis.flushdb();
    await db.delete(schema.chiiPrsnComments).where(op.eq(schema.chiiPrsnComments.mid, 1));
  });

  test('should get person comments', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/persons/1/comments',
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should create person comment', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'post',
      url: '/persons/1/comments',
      payload: { content: '恭喜恭喜', turnstileToken: 'fake-response' },
    });
    expect(res.statusCode).toBe(200);
    const pstID: number = res.json().id;
    const [pst] = await db
      .select()
      .from(schema.chiiPrsnComments)
      .where(op.eq(schema.chiiPrsnComments.id, pstID));
    expect(pst?.content).toBe('恭喜恭喜');
  });

  test('should not allow create person comment', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'post',
      url: '/persons/1/comments',
      payload: { content: '恭喜恭喜', turnstileToken: 'fake-response' },
    });
    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });
});
