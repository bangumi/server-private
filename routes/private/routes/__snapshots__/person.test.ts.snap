// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`person > should get person 1`] = `
Object {
  "career": Array [
    "artist",
    "seiyu",
  ],
  "collects": 841,
  "comment": 110,
  "id": 1,
  "images": Object {
    "grid": "https://lain.bgm.tv/pic/crt/g/a6/e8/1_prsn_ZdFfp.jpg?r=1597241889",
    "large": "https://lain.bgm.tv/pic/crt/l/a6/e8/1_prsn_ZdFfp.jpg?r=1597241889",
    "medium": "https://lain.bgm.tv/pic/crt/m/a6/e8/1_prsn_ZdFfp.jpg?r=1597241889",
    "small": "https://lain.bgm.tv/pic/crt/s/a6/e8/1_prsn_ZdFfp.jpg?r=1597241889",
  },
  "info": "性别 女 / 生日 1980年1月21日 / 血型 O型 / 身高 153cm",
  "infobox": Array [
    Object {
      "key": "简体中文名",
      "values": Array [
        Object {
          "v": "水树奈奈",
        },
      ],
    },
    Object {
      "key": "别名",
      "values": Array [
        Object {
          "k": "第二中文名",
          "v": "",
        },
        Object {
          "k": "英文名",
          "v": "",
        },
        Object {
          "k": "日文名",
          "v": "近藤奈々 (こんどう なな)",
        },
        Object {
          "k": "纯假名",
          "v": "みずき なな",
        },
        Object {
          "k": "罗马字",
          "v": "Mizuki Nana",
        },
        Object {
          "k": "昵称",
          "v": "奈々ちゃん、奈々さん、奈々様、お奈々、ヘッド",
        },
        Object {
          "k": "其他名义",
          "v": "",
        },
      ],
    },
    Object {
      "key": "性别",
      "values": Array [
        Object {
          "v": "女",
        },
      ],
    },
    Object {
      "key": "生日",
      "values": Array [
        Object {
          "v": "1980年1月21日",
        },
      ],
    },
    Object {
      "key": "血型",
      "values": Array [
        Object {
          "v": "O型",
        },
      ],
    },
    Object {
      "key": "身高",
      "values": Array [
        Object {
          "v": "153cm",
        },
      ],
    },
    Object {
      "key": "体重",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "BWH",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "引用来源",
      "values": Array [
        Object {
          "v": "zh.moegirl.org.cn",
        },
      ],
    },
    Object {
      "key": "星座",
      "values": Array [
        Object {
          "v": "水瓶座",
        },
      ],
    },
    Object {
      "key": "出身地区",
      "values": Array [
        Object {
          "v": "爱媛县 新居滨市",
        },
      ],
    },
    Object {
      "key": "所属公司",
      "values": Array [
        Object {
          "v": "StarCrew(事务所) / KING AMUSEMENT CREATIVE(唱片公司)",
        },
      ],
    },
    Object {
      "key": "个人状态",
      "values": Array [
        Object {
          "v": "已婚",
        },
      ],
    },
    Object {
      "key": "官网",
      "values": Array [
        Object {
          "v": "https://www.mizukinana.jp",
        },
      ],
    },
    Object {
      "key": "FanClub",
      "values": Array [
        Object {
          "v": "https://fanclub.mizukinana.jp",
        },
      ],
    },
    Object {
      "key": "Twitter",
      "values": Array [
        Object {
          "v": "https://twitter.com/NM_NANAPARTY",
        },
      ],
    },
  ],
  "lock": false,
  "name": "水樹奈々",
  "nameCN": "水树奈奈",
  "nsfw": false,
  "redirect": 0,
  "summary": "原名 近藤 奈々（こんどう なな），日本女性声优兼歌手。有个妹妹名字是近藤美香，为Daisy×Daisy主唱。

简介
自小受业余经营歌谣教室的父母影响，希望成为演歌歌手，从5岁起学习唱歌。小学和初中时经常在四国地区的各种歌唱比赛获奖，中一时以本名名义灌录单曲《杜樱》（つがざくら），中三时获“濑户内卡拉OK歌谣赛十周年记念全国大赛”总冠军，获其中一名评判赏识，邀请她初中毕业后往东京发展。高二（1997年）时入读代代木动画学院声优科，一边上学一边接受声优训练，不久她所属的事务所倒闭，几个月后加入由她的歌唱老师所成立的“产光音乐”事务所。其后参演PlayStation游戏NOёL ~La neige~，正式出道。因本名的日语读音跟另一位歌手近藤名奈完全一样，出道时以水树奈奈作为艺名。
高中就读堀越高等学校，跟堂本刚、友坂理惠以及山口纱弥加等同班。毕业后获颁了“堀越赏”，至今只有她和草彅刚以艺能科毕业而获得该奖赏。
水树奈奈原以演歌歌手为目标，自小受父亲训练演歌，高中时就读堀越高等学校中容许较低出席率的艺能科，尽管有所属的经纪公司，却迟迟未有歌手出道机会。
高中1年级时，事务所向水树奈奈提出进行配音工作的建议，于是，接受该建议的水树奈奈在高中二年级开始入读了代代木动画学院声优科，在此接受配音演员的训练。在1997年高三时，终于获得出道机会，获选PS游戏软件‘NOeL 〜La neige〜’的门仓千纱都一角，以艺名水树奈奈完成声优出道。高中毕业后，在吉祥寺的美体沙龙做着一份时薪850日元的兼职工作。
20岁时加盟King Records，2000年12月6日以单曲“想い”歌手出道，至今共推出34张单曲、12张大碟和3张精选集，歌手出道时的销情惨淡，但是之后就稳步上扬，销量不俗，惟近年因为实体唱片市场衰落以及日本个人女歌手热潮退却而出现下跌趋势。直到现时为止，音乐作品总销量约350万，声优唱片总销量排第三。演唱会场场爆满，音乐活动之外，也常接旁白以及配音的工作。
2000年开始每年冬季举行演唱会，至2003年增至每年夏、冬各一次日本巡回演出。2005、06年两度在日本武道馆举行演唱会，是继椎名碧流后第二位踏足这舞台的声优；2007、08年更先后成为首位在横滨体育馆、埼玉超级体育馆及国立代代木竞技场举行个人演唱会的声优。
2004年因配音人气动漫魔法少女奈叶中菲特·泰斯塔罗沙一角并演唱其中多首主题曲而走红，水树奈奈曾于出席发表会时透露配音过最喜欢的动漫角色为菲特·泰斯塔罗沙。
2009年7月5日在西武巨蛋举办了个人（也是声优界）最大场的演唱会，全长4小时半，曲目29首，不但刷新个人纪录，也刷新了声优界的纪录。
2009年12月31日获选第60回NHK红白歌合战，成为NHK红白歌合战史上第一位登上舞台的声优，被称“动画歌曲的女王”。
因唱功获赞赏和唱片销量成绩不俗，2010－14年均被邀请参加第61回－第65回NHK红白歌合战（第61回兼任红白应援队队员）。
2011年12月3日及4日首次登上东京巨蛋，是史上第八位以个人名义登上东京巨蛋、第四位连续两天开唱的日本女性歌手，而且是第一位声优能在东京巨蛋举行演唱会，再次破声优纪录。
2013年7月14日在日本巡回演唱会“NANA MIZUKI LIVE CIRCUS 2013”大阪城音乐厅场次中，宣布将于台湾举办首次海外演唱会。
2013年11月23日及24日，在台湾台北市华山1914创意文化园区台北音乐传记（Legacy Taipei）举办首次海外演唱会“NANA MIZUKI LIVE CIRCUS 2013+”。2013年11月24日，在“NANA MIZUKI LIVE CIRCUS 2013+”宣布，将于2014年再次于台湾举办海外演唱会。
2014年6月，NANA MIZUKI LIVE CIRCUS×CIRCUS＋×WINTER FESTA BD作品首周以1.8万张销售额获得公信榜音乐部门第一名，综合部门第二名且打破纪录以连续9张BD作品蝉联音乐部门榜上第一名。第二名为安室奈美惠（5张）。
2016年4月9、10日第二次在东京巨蛋举行演唱会，成为在东京巨蛋开唱过最多次的日本个人女歌手之一，并且宣布于2016年9月22日于阪神甲子园球场举办个人演唱会。

演艺经历
1998年在Pioneer LDC（现已改名为日本Geneon环球娱乐）旗下推出一张CD。
1999年10月在网络电台JAM STATION主持自己的节目《NANA CHANNEL》。
2000年1月23日于电台节目制作团队的协助下，举行第一场个人演唱会，被King Records制作人三嶋章夫所赏识，其后正式成为King Records（但不是King Records旗下制作动画、声优音乐为主的STARCHILD）的歌手。
2001年9月受现在经纪人建议，离开她在东京寄宿的歌唱老师住所与其经纪公司，转往Sigma Seven事务所。
2001年曾因《妹妹公主》而与桑谷夏子、小林由美子、望月久代组成声优组合“Prits”，并发行过3张单曲和1张大碟。
同年（2001年）再因动画《七小花》（七人のナナ）而与秋田まどか、浅木舞、中原麻衣、名冢佳织、福井裕佳梨、桃森すもも组成“nana×nana”主唱该动画的主题曲。
2003年7月7日成立官方俱乐部S.C. NANA NET。
2005年1月2日于日本武道馆举行演唱会，是继椎名碧流后的第二位以个人名义在武道馆开演唱会的声优。
2006年1月21日再次踏足日本武道馆，自2000年和2001年以来再次举行生日演唱会，演唱会当天更自弹自唱“宝物”。
2006年10月29日日本电视台TBS Channel播放特别制作的水树奈奈特集，片长约90分钟，节目中有独家访问、2006年夏季演唱会精选片段和本人亲自下厨等珍贵镜头。
2007年2月12日在横滨体育馆举行个人史上最长最大的演唱会，成为首位以个人名义使用这场地的声优。
2007年3月3日第一届“声优Awards”获歌唱赏。
2009年2月5日首度亮相全国广播的NHK电视音乐节目MUSIC JAPAN，并于同年4月开始担任节目旁白。
2009年7月5日举办个人首场巨蛋演唱会，现场人数涌进3万人，为2009年唯一使用巨蛋级场地举办演唱会的女性歌手。
2009年12月31日初次参与NHK红白歌合战。
2010年3月6日第四届“声优Awards”获富山敬赏。
2010年12月31日第二次参与NHK红白歌合战。
2011年2月6日荣获“Billboard JAPAN MUSIC AWARDS 2010年年度优秀歌手赏”。
2011年3月27日参与niconico为2011年日本本州岛海域地震举行的赈灾款募金义演，总金额达315万1800円。
2011年12月3日及4日举办个人首场东京巨蛋演唱会，为史上第八位日本女歌手以个人名义登上东京巨蛋，累计现场人数为8万人。这场演唱会还创下“东京巨蛋史上的第一场全天域星象投影表演”的纪录。
2011年12月31日第三次参与NHK红白歌合战。
2012年3月3日荣获“Billboard JAPAN MUSIC AWARDS”年度最佳动画歌手。并在2012年度蝉联该奖项。
2012年12月31日连续第四年参与NHK红白歌合战。
2013年5月15日发表第一首合唱单曲“Preserved Roses”（合唱者为T.M.Revolution）。
2013年11月10日参与于新加坡举行的Anime Festival Asia 2013: Valvrave Night演出，乃首个于国外进行的公开活动。
2013年11月23日及24日举行第一场海外演唱会“NANA MIZUKI LIVE CIRCUS 2013+”，地点为台湾Legacy Taipei，其中24日场次于日本、台湾、新加坡、香港、泰国、印尼等65间戏院进行直播。
2013年12月31日连续第五年参与NHK红白歌合战。
2014年3月13日于文部科学省文化厅举办的“2013年度（第64届）艺术选奖”中获得大众艺能部门的新人赏。
2014年6月1日 在演唱会上宣布追加海外公演，在新加坡及台湾举行，新加坡场于9月27日在“RESORTS WORLD THEATRE”演出，台湾场则于10月4、5两日在台北市立大学天母校区体育馆举行。
2014年12月31日连续第六年参与NHK红白歌合战。
2015年7月24日首度在MUSIC STATION中演出，是继椎名碧流后的第二位在该节目中演出的声优。
2016年4月9日及10日再度于东京巨蛋举行演唱会，成为东京巨蛋中演唱会举行次数最多的日本个人女性歌手之一（另一个为安室奈美惠）。
2016年9月22日首度于阪神甲子园球场举行演唱会，成为首位在阪神甲子园球场举行演唱会的个人歌手。
2017年7、8月于帝国剧场参与以卡洛尔·金为题材的音乐剧，并与平原绫香担任双主演。

个人生活
水树奈奈的父亲为近藤信光，母亲为近藤训子 。名字的由来是因为父亲是演员冈田奈奈的粉丝，因为避免本名“近藤奈奈”与歌手近藤名奈出现混淆，所以决定使用“水树奈奈”作为艺名。
她的妹妹近藤美香为Daisy×Daisy主唱，艺名为Mika。
家庭环境并不富裕，所以学生时代是学校里少数几乎全勤的学生，并为领奖学金以贴补家用而十分用功，经常学年第一。2008年10月29日，水树奈奈的父亲去世 。
持有书友会毛笔部七段、珠算三级、心算二级、自动档汽车驾驶执照等资格。
与同事务所所属，在电台节目“スマイルギャング”中共演的声优福圆美里交情深厚，每年在彼此生日临近的时候都会互相交换生日礼物。因演出“放浪男孩”而与同事务所所属声优濑户麻沙美关系亲密，经常一起吃饭，将瀬戸视作自己妹妹般。2004年于因为演出魔法少女奈叶而和声优田村由香里关系亲密。与能登麻美子和泽城美雪一起去旅行后，交情加深。
2020年7月6日，与作为音乐界从业者的丈夫登记结婚。11月6号，通过官网宣布怀孕，目前已经进入安定期，将暂时工作一段时间。
水树的父亲持有硬式棒球的裁判执照，水树则是个颇热情的阪神虎爱好者。最初比较喜欢高校棒球，受喜欢读卖巨人的父亲影响而成为了巨人的球迷。1996年，当清原和博（已在2008年球季结束后退休）由Free Agent转至巨人时，开始觉得巨人不好，而转而喜欢阪神虎直至现在。特别尊敬星野仙一。
水树为第11张专辑SMASHING ANTHEMS拍专辑照时，到过阪神虎的主场阪神甲子园球场取景。在阪神甲子园球场举行演唱会为水树的其中一个梦想，场方则以保护天然草皮为理由，每年只批准一组歌唱单位举行演唱会。每年在甲子园球场举行演唱会的TUBE宣布2016年不会在此场举行演唱会，结果水树得到这个机会。

其他
2006年，井上喜久子曾经希望水树能够加入17岁教。
2008年6月19日，21岁的送报员堀大树用手机在2ch的声优板块预告中模仿加藤智大，发出拟于7月5日-6日在代代木体育馆举行的LIVE FIGHTER 2008 BLUE SIDE/RED SIDE中，杀害水树奈奈的犯罪预告，后嫌疑人被警方监控、逮捕。
2008年10月29日晚间8时40分，水树之父近藤信光因多器官衰竭去世，卒年75岁。
2009年11月21日于东京国际展示场举行的俱乐部活动中挑战吉尼斯世界纪录，原本要挑战三个项目，有两个项目挑战成功，分别是7014人全场手牵手形成最长人浪，以及动员6961人创下最多人于同一时间一起吹party blowers，其后由吉尼斯世界纪录职员即场颁发证书。
至于无法挑战成功的项目，就是于室内发出最大音量的人声，现时纪录由香港饶舌团体农夫举行的演唱会保持。
2009年12月13日，终于与一直想见面的同乡（爱媛县出身）・演员真锅香织在剧场版动画“雷顿教授与永远的歌姬”的首映会上见面。
2010年2月20日，被任命为新居浜观光大使。
2010年10月，台湾金牌大风取得代理并发售IMPACT EXCITER台压盘，是水树的第一张海外盘。
2011年1月21日，于生日当天发售个人第一本自传“深爱”。
2012年2月，中国大陆星外星唱片取得代理并发售IMPACT EXCITER引进版，是水树正式在大陆发行的首张个人专辑。
2012年5月，台湾金牌大风取得代理并发售NANA MIZUKI LIVE CASTLE×JOURNEY台压盘，是第一张水树的演唱会海外盘。
2012年4月17日，再次被任命为伊予观光大使。
2012年6月10日播出的新堂本兄弟中与高中同学堂本刚实现了久隔14年见面的会谈，称水树为近藤さん，并在业界传开来。
2012年12月5日，在富士电视台的FNS歌谣祭中水树与堂本刚所在的团队KinKi Kids坐在同一个圆桌时，并从刚君处听说他在他的演唱会上提到了水树。
2014年1月16日，发售文库版本之自传。
2014年6月10日，25岁男子福元一辉为了抽奖在演唱会与水树见面而购买1000包卡乐B薯片，将每包薯片上的抽奖券取去后丢弃在树林中，因违反《废弃物处理法》而被警方逮捕。
2014年6月22日，由于水树的声带发炎导致失声，为日后着想需要休息一段时间，原定于当日以及6月28日和29日举行的巡回公演被迫取消。

YouTube频道：
https://www.youtube.com/user/mizukinanaKING",
  "type": 1,
}
`;

exports[`person > should get person casts 1`] = `
Object {
  "data": Array [],
  "total": 0,
}
`;

exports[`person > should get person collects 1`] = `
Object {
  "data": Array [],
  "total": 0,
}
`;

exports[`person > should get person works 1`] = `
Object {
  "data": Array [
    Object {
      "positions": Array [
        Object {
          "appearEps": "",
          "summary": "",
          "type": Object {
            "cn": "开发",
            "en": "Developer",
            "id": 1001,
            "jp": "開発元",
          },
        },
      ],
      "subject": Object {
        "id": 4,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/a8/7f/4_cMMK5.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/a8/7f/4_cMMK5.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/a8/7f/4_cMMK5.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/a8/7f/4_cMMK5.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/a8/7f/4_cMMK5.jpg",
        },
        "info": "2008-07-17 / NDS / ACT / SNKプレイモア",
        "locked": false,
        "name": "メタルスラッグ7",
        "nameCN": "合金弹头7",
        "nsfw": false,
        "rating": Object {
          "count": Array [
            0,
            0,
            0,
            1,
            8,
            47,
            43,
            39,
            3,
            2,
          ],
          "rank": 3560,
          "score": 6.9,
          "total": 143,
        },
        "type": 4,
      },
    },
    Object {
      "positions": Array [
        Object {
          "appearEps": "",
          "summary": "",
          "type": Object {
            "cn": "游戏设计师",
            "en": "Game Designer",
            "id": 1003,
            "jp": "ゲームクリエイター",
          },
        },
      ],
      "subject": Object {
        "id": 6,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/16/79/6_gC3iJ.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/16/79/6_gC3iJ.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/16/79/6_gC3iJ.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/16/79/6_gC3iJ.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/16/79/6_gC3iJ.jpg",
        },
        "info": "2007年10月10日 / Xbox 360、PS3、PC / FPS / Valve Corporation",
        "locked": false,
        "name": "Team Fortress 2",
        "nameCN": "军团要塞2",
        "nsfw": false,
        "rating": Object {
          "count": Array [
            0,
            0,
            0,
            1,
            3,
            21,
            60,
            53,
            23,
            11,
          ],
          "rank": 1360,
          "score": 7.59,
          "total": 172,
        },
        "type": 4,
      },
    },
  ],
  "total": 247,
}
`;

exports[`person comments > should get person comments 1`] = `
Array [
  Object {
    "content": "7月7日 77结婚",
    "createdAt": 1400517144,
    "creatorID": 287622,
    "id": 12345670,
    "mainID": 1,
    "relatedID": 0,
    "replies": Array [
      Object {
        "content": "是6号结的",
        "createdAt": 1400517144,
        "creatorID": 287622,
        "id": 12345671,
        "mainID": 1,
        "relatedID": 12345670,
        "state": 0,
        "user": Object {
          "avatar": Object {
            "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
            "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
            "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
          },
          "group": 0,
          "id": 287622,
          "joinedAt": 0,
          "nickname": "nickname 287622",
          "sign": "sing 287622",
          "username": "287622",
        },
      },
    ],
    "state": 0,
    "user": Object {
      "avatar": Object {
        "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
        "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
        "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
      },
      "group": 0,
      "id": 287622,
      "joinedAt": 0,
      "nickname": "nickname 287622",
      "sign": "sing 287622",
      "username": "287622",
    },
  },
  Object {
    "content": "生日快乐！",
    "createdAt": 1400517144,
    "creatorID": 287622,
    "id": 12345672,
    "mainID": 1,
    "relatedID": 0,
    "replies": Array [],
    "state": 0,
    "user": Object {
      "avatar": Object {
        "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
        "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
        "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
      },
      "group": 0,
      "id": 287622,
      "joinedAt": 0,
      "nickname": "nickname 287622",
      "sign": "sing 287622",
      "username": "287622",
    },
  },
]
`;

exports[`person comments > should not allow create person comment 1`] = `
Object {
  "code": "NEED_LOGIN",
  "error": "Unauthorized",
  "message": "you need to login before creating a comment",
  "statusCode": 401,
}
`;
